### ISSUE TITLE
Make attribute-indentation rule "fixable"

### ISSUE THREAD

**Issue by tstewart15** [ember-template-lint/ember-template-lint#1523]
Currently, the attribute-indentation rule does not implement a "fix" code path, which would enable use of "--fix" to autofix attribute-indentation violations.

**Comment by tstewart15** [comment]
I would like to work on this task in the coming weeks, FYI.

**Comment by dcyriller** [comment]
In case: there is a [how-to](https://github.com/ember-template-lint/ember-template-lint/blob/master/docs/fixer.md) to help you implement the fixer.

**Comment by tstewart15** [comment]
Thanks @dcyriller ! I'm also wondering if there is any good documentation on how to use ember-template-recast? This is my first time using it, and there isn't much I can find by looking around https://github.com/ember-template-lint/ember-template-recast , especially around manipulating the whitespace (like I'll have to do for this issue).

**Comment by dcyriller** [comment]
template-recast’s README documents the package API (mainly the parse / print / transform functions).

If you’d like discover the ast, what the different types are and so on, you can play with https://astexplorer.net

If you want to have a more precise idea of Handlebars ast, you can look at [Handelbars ts interfaces](https://github.com/glimmerjs/glimmer-vm/blob/v0.61.1/packages/@glimmer/syntax/lib/types/handlebars-ast.ts) in @glimmer/syntax.

At some point you might want to create nodes. You’ll have to use @glimmer/syntax builders. Their ts interfaces [are here](https://github.com/glimmerjs/glimmer-vm/blob/v0.61.1/packages/%40glimmer/syntax/lib/builders.ts).

Some other people may have other ideas / resources but this is what comes to my mind.

**Comment by Alonski** [comment]
@tstewart15 @dcyriller There is also this "workshop" by @Turbo87
https://github.com/simplabs/ast-workshop

**Comment by Turbo87** [comment]
just to be clear, this issue is rather hard to fix via ember-template-recast, because the regular AST does not contain whitespace information.

I guess something like https://github.com/ember-template-lint/ember-template-lint-plugin-prettier/blob/master/lib/rules/prettier.js#L103 might be needed to solve this, after working on the string level and adjusting the line indentations there.

**Comment by Alonski** [comment]
Would it be possible to add whitespace info to the AST?

**Comment by Turbo87** [comment]
yes, but we shouldn't. AST are not meant to have that kind of information, because they are **abstract** syntax trees.

**Comment by Alonski** [comment]
Hmm ok sort of makes sense 👍

**Comment by rwjblue** [comment]
> just to be clear, this issue is rather hard to fix via ember-template-recast, because the regular AST does not contain whitespace information.

Confirm.

> yes, but we shouldn't. AST are not meant to have that kind of information, because they are abstract syntax trees.

Yep, we kinda want a CST (concrete syntax tree), but alas that isn't what we have available to us.

**Comment by Alonski** [comment]
@rwjblue Is this just because of how Glimmer works?

**Comment by tstewart15** [comment]
How should we proceed with issue? Close it in favor of pending [Prettier support](https://github.com/ember-template-lint/ember-template-lint-plugin-prettier)?

**Comment by rwjblue** [comment]
> just to be clear, this issue is rather hard to fix via ember-template-recast, because the regular AST does not contain whitespace information.

I do think that it could be done... 🤔 

I took a stab at demonstrating the basic technique that _should_ allow for this to work over in https://github.com/ember-template-lint/ember-template-recast/pull/362. I'll likely land that test, but it'll most likely be up to someone with time to actually migrate that into a fixer here...

**Comment by VincentMolinie** [comment]
Thanks @rwjblue for your example it helped me a lot. I did an attempt, any comment / help is welcome: #2271

**Comment by lifeart** [comment]
Closing issue, but it's still listed here: https://github.com/ember-template-lint/ember-template-lint/issues/3209
